### DISCUSSION TITLE
Different Postcode returned for GB in self hosted vs public nominatim installation

### DISCUSSION THREAD

**Discussion by patway211** [osm-search/Nominatim#3570] [Q&A]
We are getting different UK postcode for many  Lat,Long pair as compared to nominatim.opentreetmap in a server with fresh installation of nominatim full planet. Updated Database and GB postcode.

Followed all steps as mentioned @
https://nominatim.org/release-docs/latest/admin/Install-on-Ubuntu-22/
https://nominatim.org/release-docs/latest/admin/Import/

Got the following two before start of import to project directory :
wget https://www.nominatim.org/data/gb_postcode_data.sql.gz
wget https://www.nominatim.org/data/us_postcode_data.sql.gz

Software Environment :
    • Nominatim version: v4.5.0
    • Postgresql version: 14
    • Postgis version: 3.2.0
    • OS: Ubuntu Ubuntu 22.04.4 LTS
Hardware Configuration :
    • RAM: 128GB
    • number of CPUs: 16
    • type and size of disks: 1.5TB SSD
Nominatim Configuration:
NOMINATIM_FLATNODE_FILE=/mnt/geo_lon1_04/flatfile/flatnode.file

Here is an example :
Lat : 53.4367566
Long -1.3470283

In nominatim.openstreetmap 
https://nominatim.openstreetmap.org/ui/reverse.html?lat=53.4367566&lon=-1.3470283&zoom=18
Selwyn Street, Eastwood, Rawmarsh, Rotherham, South Yorkshire, England, 
S65 1QN, 
United Kingdom 

In Self hosted : 
Selwyn Street, Eastwood, Rawmarsh, Rotherham, South Yorkshire, England, 
S65 1QW,
United Kingdom

The post code expected : S65 1QN

I am new to Nominatim. 
Did I miss any step ? How to debug and/or resolve this ?

**Comment by mtmail** [comment]
`S65 1QN` look indeed closer

Can you post the output of the following database query? (indexed_status=0 means it got indexed).
```sql
select postcode, ST_AsText(geometry), indexed_date, indexed_status from location_postcode where country_code='gb' and postcode like 'S65 1Q%';
```

![image](https://github.com/user-attachments/assets/c3ddbc70-183a-4b12-a832-4a926881df21)

**Comment by patway211** [comment]
Here is the output of : select postcode, ST_AsText(geometry), indexed_date, indexed_status from location_postcode where country_code='gb' and postcode like 'S65 1Q%';

postcode |          st_astext                          |        indexed_date                | indexed_status 
----------+------------------------------+----------------------------+----------------
 S65 1QB  | POINT(-1.34246 53.43762)     | 2024-10-28 22:51:16.022757 |              0
 S65 1QD  | POINT(-1.34652 53.4337)      | 2024-10-28 22:51:16.022757 |              0
 S65 1QE  | POINT(-1.3416999 53.43775)   | 2024-10-28 22:51:16.022757 |              0
 S65 1QJ  | POINT(-1.3475 53.43664)      | 2024-10-28 22:51:16.022757 |              0
 S65 1QL  | POINT(-1.34767 53.43709)     | 2024-10-28 22:51:16.022757 |              0
 S65 1QN  | POINT(-1.34683 53.43716)     | 2024-10-28 22:51:16.022757 |              0
 S65 1QP  | POINT(-1.3462099 53.43754)   | 2024-10-28 22:51:16.022757 |              0
 S65 1QR  | POINT(-1.34688 53.43841)     | 2024-10-28 22:51:16.022757 |              0
 S65 1QS  | POINT(-1.34616 53.4382)      | 2024-10-28 22:51:16.022757 |              0
 S65 1QT  | POINT(-1.34571 53.43751)     | 2024-10-28 22:51:16.022757 |              0
 S65 1QU  | POINT(-1.34314 53.43837)     | 2024-10-28 22:51:16.022757 |              0
 S65 1QW  | POINT(-1.34711 53.43773)     | 2024-10-28 22:51:16.022757 |              0
 S65 1QX  | POINT(-1.34378 53.43848)     | 2024-10-28 22:51:16.084834 |              0
 S65 1QY  | POINT(-1.3415 53.43916)      | 2024-10-28 22:51:16.084834 |              0
 S65 1QA  | POINT(-1.3449581 53.4373175) | 2024-10-26 16:00:54.881916 |              0
(15 rows)

Output of check-database : 

2024-10-30 16:05:11: Using project directory: /srv/nominatim/nominatim-project
2024-10-30 16:05:11: Checking database
Checking database connection ... OK
Checking database_version matches Nominatim software version ... OK
Checking for placex table ... OK
Checking for placex content ... OK
Checking that tokenizer works ... OK
Checking for wikipedia/wikidata data ... OK
Checking indexing status ... OK
Checking that database indexes are complete ... OK
Checking that all database indexes are valid ... OK
Checking TIGER external data table. ... not applicable

**Reply by mtmail** [reply]
nominatim.openstreetmap.org was installed with Nominatim 4.4 (or maybe 4.3 even). It's possible that reindexing the area or reimporting on those servers will show the same result. I created https://github.com/osm-search/Nominatim/issues/3572 based on your question. Your installation looks alright (apart from the postcodes) and I don't think you can further debug or fix it locally.

**Reply by patway211** [reply]
Thank you. We tried with the associated code changes in #3572 and #3575. We are getting better results than before.

**Comment by lonvia** [comment]
I've added some technical details to the #3572.

Here I just want to add that I would consider both postcodes perfectly good answers in this situation. Reverse doesn't compute the nearest postcode for you, it returns the closest OSM object. What you see in the response is the postcode of "Selwyn Street". Given the small radius of UK postcodes, there are usually multiple postcodes that belong to one street. Nominatim will just give you one of them.

**Reply by patway211** [reply]
Thank you for the details. We re-imported the data with updated code and getting better results than before for the sample data we tried.
